google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java

```diff
@@ -0,0 +1,347 @@
+  }
+
+  @Override
+  public boolean isModified() {
+    return !yaml.getText().equals(deploymentConfiguration.getYamlPath())
+        || !dockerfile.getText().equals(deploymentConfiguration.getDockerFilePath());
+  }
+
+  @Override
+  public void reset() {
+    yaml.setText(deploymentConfiguration.getYamlPath());
+    dockerfile.setText(deploymentConfiguration.getDockerFilePath());
+
+    toggleDockerfileSection();
+  }
+
+  @Override
+  public void apply() throws ConfigurationException {
+    validateConfiguration();
+    if (!isValidConfigurationFile(yaml.getText())) {
+      throw new ConfigurationException(
+          GctBundle.getString("appengine.deployment.error.staging.yaml"));
+    }
+
+    if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
+      throw new ConfigurationException(
+          GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+    }
+
+    deploymentConfiguration.setYamlPath(yaml.getText());
+    deploymentConfiguration.setDockerFilePath(dockerfile.getText());
+
+    toggleDockerfileSection();
+  }
+
+  @Override
+  public void disposeUIResources() {
+    // Do nothing.
+  }
+
+  @Nls
+  @Override
+  public String getDisplayName() {
+    return GctBundle.getString("appengine.flexible.facet.name");
+  }
+
+  private boolean isRuntimeCustom() {
+    return APP_ENGINE_PROJECT_SERVICE.getFlexibleRuntimeFromAppYaml(yaml.getText())
+        == FlexibleRuntime.CUSTOM;
+  }
+
+  /**
+   * Tests if a file is valid by checking whether it exists, is a regular file and is well formed.
+   */
+  private boolean isValidConfigurationFile(String path) {
+    try {
+      if (!Files.exists(Paths.get(path))
+          || !Files.isRegularFile(Paths.get(path))) {
+        return false;
+      }
+    } catch (InvalidPathException ipe) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Validates the configuration and turns on/off any necessary warnings.
+   */
+  private void validateConfiguration() {
+    boolean showError = false;
+
+    if (!isValidConfigurationFile(yaml.getText())) {
+      errorIcon.setVisible(true);
+      errorMessage.setVisible(true);
+      errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.yaml"));
+      showError = true;
+    }
+
+    if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
+      errorIcon.setVisible(true);
+      errorMessage.setVisible(true);
+      errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+      showError = true;
+    }
+
+    errorIcon.setVisible(showError);
+    errorMessage.setVisible(showError);
+  }
+
+  private void toggleDockerfileSection() {
+    boolean enabled = isRuntimeCustom();
+    dockerfile.setEnabled(enabled);
+    genDockerfileButton.setEnabled(enabled);
+    // Shows no Dockerfile label if runtime is Java and yaml is valid.
+    noDockerfileLabel.setVisible(!enabled && isValidConfigurationFile(yaml.getText()));
```

Comment:
If the YAML file doesn't exist, enabled will be false (because runtime will be UNKNOWN) and the noDockerfile message will show up saying that runtime is set to "java" in the YAML file. The second part of the and aims at hiding the message in that case.